### PR TITLE
(WIP) Implementation of structural types

### PR DESCRIPTION
**Pull request by LukasKellenberger**

### CODE REVIEW

File: nir/src/main/scala/scala/scalanative/nir/Attrs.scala [hunk 1]
```diff
@@ -12,6 +12,8 @@ object Attr {
   final case object NoInline     extends Inline // should never inline
   final case object AlwaysInline extends Inline // should always inline
 
+  final case object StructDisp   extends Attr //attr for defs that can be used for structural dispatch
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Attrs.scala]: `Dyn`

Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Attrs.scala]: Also needs update to NIR docs.

File: nir/src/main/scala/scala/scalanative/nir/Attrs.scala [hunk 2]
```diff
@@ -21,11 +23,13 @@ object Attr {
   sealed abstract class Pin                         extends Attr
   final case class PinAlways(dep: Global)           extends Pin
   final case class PinIf(dep: Global, cond: Global) extends Pin
+  final case class WeakPin(dep: Global)             extends Pin
 }
 
 final case class Attrs(inline: Inline = MayInline,
                        isPure: Boolean = false,
                        isExtern: Boolean = false,
+                       usedForStructDisp: Boolean = false,
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Attrs.scala]: `isDyn`

File: nir/src/main/scala/scala/scalanative/nir/Ops.scala [hunk 3]
```diff
@@ -50,6 +51,7 @@ object Op {
   final case class Classalloc(name: Global)                        extends Op
   final case class Field(ty: Type, obj: Val, name: Global)         extends Op
   final case class Method(ty: Type, obj: Val, name: Global)        extends Op
+  final case class DynMethod(obj: Val, sign: String)               extends Op
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Ops.scala]: `signature`

File: nir/src/main/scala/scala/scalanative/nir/Shows.scala [hunk 4]
```diff
@@ -32,13 +32,16 @@ object Shows {
     case Attr.NoInline     => "noinline"
     case Attr.AlwaysInline => "alwaysinline"
 
+    case Attr.StructDisp   => "structdisp"
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Shows.scala]: `"dyn"`

File: nir/src/main/scala/scala/scalanative/nir/Attrs.scala [hunk 5]
```diff
@@ -21,11 +23,13 @@ object Attr {
   sealed abstract class Pin                         extends Attr
   final case class PinAlways(dep: Global)           extends Pin
   final case class PinIf(dep: Global, cond: Global) extends Pin
+  final case class WeakPin(dep: Global)             extends Pin
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Attrs.scala]: `Pin`, `CondPin`, `WeakPin`.

File: nir/src/main/scala/scala/scalanative/nir/Ops.scala (hunk 3 above)
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Ops.scala]: `Dynmethod`

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala [hunk 6]
```diff
@@ -21,33 +22,43 @@ final class BinaryDeserializer(_buffer: => ByteBuffer) {
...
     this.deps = null
     map
   }
 
-  private var deps: mutable.Set[Dep]        = _
-  private var links: mutable.Set[Attr.Link] = _
+  private var deps:  mutable.Set[Dep]           = _
+  private var links: mutable.Set[Attr.Link]     = _
+  private var weaks: mutable.Set[Attr.WeakPin]  = _
+  private var dyns:   mutable.Set[String]       = _
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: Need to merge `deps` and `weaks` into just `deps`.

File: tools/src/main/scala/scala/scalanative/linker/Linker.scala [hunk 7]
```diff
@@ -57,11 +71,21 @@ final class Linker(dotpath: Option[String], paths: Seq[String]) {
...
+              (dynMethods(weak.toSeq ++ newWeaks, newDyns) ++ dynMethods(newWeaks, dyn.toSeq)).foreach {
+                dep =>
+                  writeEdge(dep._1, dep._2)
+                  direct.push(dep._2)
+                  structuralMethods += dep._2
+              }
+
+              weak ++= newWeaks
+              dyn ++= newDyns
+
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/linker/Linker.scala]: Move it to `deps.foreach { ... case Dep.Weak(signature) => ... }`

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 8]
```diff
@@ -848,8 +855,28 @@ abstract class NirCodeGen
...
+    def genDynMethodCall(sym: Symbol, self: Val, argsp: Seq[Tree], focus: Focus): Focus = {
+
+      val sig          = genMethodSig(sym)
+      val methodName   = genMethodName(sym).id
+      val (args, last) = genMethodArgs(sym, argsp, focus)
+
+      val method = last withOp Op.DynMethod(self, methodName.toString)
+      val values = self +: args
+
+      //unsupported(methodName)
```
Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Dead comment?

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 9]
```diff
@@ -848,8 +855,28 @@ abstract class NirCodeGen
...
+      method withOp Op.Call(sig, method.value, values)
+    }
```
Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: Needs more tests.

**Comment by densh** [comment]
@LukasKellenberger can you please sign [Scala CLA](https://www.lightbend.com/contribute/cla/scala) ?

**Comment by LukasKellenberger** [comment]
I signed the Scala CLA.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala [hunk 10]
```diff
@@ -848,8 +855,37 @@ abstract class NirCodeGen
...
+      val methodName   = genSignature(genMethodName(sym))
+      val (args, last) = genMethodArgs(sym, argsp, focus)
+
+      val method = last withOp Op.Dynmethod(self, methodName.toString)
+      val values = self +: args
+
+      method withOp Op.Call(sig, method.value, values)
+    }
+
+    def genSignature(methodName: nir.Global): String = {
```
Comment (LukasKellenberger) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala]: This method is duplicated several time, where should I put it ?

File: tools/src/main/scala/scala/scalanative/compiler/analysis/ClassHierarchy.scala [hunk 11]
```diff
@@ -81,16 +81,81 @@ object ClassHierarchy {
...
+      case (hash, c) =>
+        val inter = hash ^ c.toInt
+        inter + (inter << 1) + (inter << 4) + (inter << 5) + (inter << 7) + (inter << 8) + (inter << 40);
+    }
 
-    lazy val typeValue: Val.Struct = Val
-      .Struct(Global.None, Seq(Val.I32(id), Val.String(name.id), vtableValue))
+    lazy val dynDispatchTableValue: Val = {
+      // If there a value for null pointer ?
+      lazy val defaultPtr: Val = allmethods.head.value
```
Comment (LukasKellenberger) [inline tools/src/main/scala/scala/scalanative/compiler/analysis/ClassHierarchy.scala]: Is there a value for "null" pointers ? (When there are no dynamic methods)

**Comment by densh** [comment]
@LukasKellenberger As discussed today this needs to be split into a number of smaller PRs.

**Pull request closed without merging**